src/charm.py

```diff
@@ -7,29 +7,176 @@
+        """
+        # TODO: Can we make this generic and share with other charms?
+        name = "kiali"
+        if not self._container.can_connect():
+            LOGGER.warning(f"Container is not ready, cannot configure {name}")
+            return
+
+        layer = self._generate_kiali_layer()
+        try:
+            new_config = yaml.dump(self._generate_kiali_config())
+        except GrafanaSourceError as e:
+            LOGGER.warning(f"Failed to generate {name} configuration, got error: {e}")
+            # TODO: actually shut down the service and remove the configuration
+            # LOGGER.warning(f"Shutting down {name} service and removing existing configuration")
+            return
+
+        should_restart = not _is_container_file_equal_to(
+            self._container, str(KIALI_CONFIG_PATH), new_config
+        )
+        self._container.push(KIALI_CONFIG_PATH, new_config, make_dirs=True)
+        self._container.add_layer(name, layer, combine=True)
+        self._container.autostart()
+
+        if should_restart:
+            LOGGER.info(f"new config detected for {name}, restarting the service")
+            self._container.replan()
+
+    def _generate_kiali_config(self) -> dict:
+        """Generate the Kiali configuration."""
+        prometheus_url = self._get_prometheus_source_url()
+        return {
+            "auth": {
+                "strategy": "anonymous",
+            },
+            "external_services": {"prometheus": {"url": prometheus_url}},
+            # TODO: Use the actual istio namespace
+            "istio_namespace": "istio-system",
+            "server": {"port": KIALI_PORT, "web_root": "/kiali"},
+        }
+
+    @staticmethod
+    def _generate_kiali_layer() -> Layer:
+        """Generate the Kiali layer."""
+        # TODO: Add pebble checks?
+        return Layer(
+            {
+                "summary": "Kiali",
+                "description": "The Kiali dashboard for Istio",
+                "services": {
+                    "kiali": {
+                        "override": "replace",
+                        "summary": "kiali",
+                        "command": f"/opt/kiali/kiali -config {KIALI_CONFIG_PATH}",
+                        "startup": "enabled",
+                    }
+                },
+            }
+        )
+
+    def _get_prometheus_source_url(self):
+        """Get the Prometheus source configuration.
+
+        Raises a SourceNotAvailableError if there are no sources or the data is not complete.
+        """
+        if not (prometheus_sources := self._prometheus_source.sources):
+            raise GrafanaSourceError("No Prometheus sources available")
+        if len(prometheus_sources) > 1:
+            raise GrafanaSourceError("Multiple Prometheus sources available, expected only one")
+        if not (url := prometheus_sources[0].get("url", None)):
+            raise GrafanaSourceError("Prometheus source data is incomplete - url not available")
```

Comment:
Why is this exception called `GrafanaSourceError` when it is relating to a Prometheus source.

Reply:
Yeah agreed, this is silly.  Its because its coming from the Grafana source datasource, but the user doesn't care about that.  I've changed it to Prometheus